crates/spfs-cli/cmd-fuse/src/cmd_fuse.rs

```diff
@@ -268,11 +289,40 @@ impl CmdFuse {
                 _ = terminate.recv() => Err(miette!("Terminate signal received, filesystem shutting down")),
                 _ = interrupt.recv() => Err(miette!("Interrupt signal received, filesystem shutting down")),
                 _ = quit.recv() => Err(miette!("Quit signal received, filesystem shutting down")),
+                _ = heartbeat_recv.recv() => {
+                    heartbeat_failed = true;
+                    Err(miette!("Heartbeat monitor triggered, filesystem shutting down"))
+                }
             };
-            // the filesystem task must be fully terminated in order for the subsequent unmount
+
+            heartbeat_monitor.abort_handle().abort();
+            if let Err(err) = heartbeat_monitor.await {
+                tracing::warn!("Heartbeat monitor failed: {err:?}");
+            }
+
+            // The filesystem task must be fully terminated in order for the subsequent unmount
             // process to function. Otherwise, the background task will keep this process alive
             // forever.
-            abort_handle.abort();
+            //
+            // Exception: if spfs-monitor died without cleaning up the runtime,
+            // fuser::Session::run will not exit and there's no way to tell it
+            // to break out of its loop. When the fuse filesystem is included
+            // in the lowerdir of an overlayfs, even though `fusermount -u`
+            // appears to succeed, and the mount disappears from /etc/mtab,
+            // spfs-fuse stays alive and still functions when accessed via the
+            // overlayfs. As a last-ditch effort to avoid leaving a spfs-fuse
+            // process running forever, we will return without attempting any
+            // cleanup.
+            if heartbeat_failed {
+                return res;
+            }
+            if !join_handle.is_finished() {
+                // XXX: Calling unmount has no apparent effect!
+                unmount_callable.lock().unwrap().unmount().into_diagnostic().wrap_err("FUSE unmount failed")?;
+                tracing::trace!("Joining FUSE session");
+                join_handle.await.into_diagnostic().wrap_err("FUSE join_handle await failed")?.into_diagnostic().wrap_err("FUSE session failed after unmount")?;
+                tracing::trace!("FUSE session joined");
```

Comment:
This is meant to shutdown cleanly, but in practice it will park forever waiting on the join_handle. In order for the run loop to exit, the fuse filesystem needs to be unmounted, but the unmount call here doesn't work. As mentioned in the comment, even fusermount can apparently succeed at unmounting the path, but the spfs filesystem continues to function normally when it is part of an overlayfs stack.

On the other hand, just returning and allowing the process to terminate will allow the fusermount auto_unmount to kick in. At least in my testing, after `kill -9`'ing spfs-monitor and allowing the heartbeat to expire, spfs-fuse exits and the accompanying fusermount process goes away.

Attempting to access /spfs after that point does this:
```
jrray@STRANDED:~/git/spk$ ls /spfs
ls: reading directory '/spfs': Transport endpoint is not connected
```

(This is in my ubuntu home environment.)

Reply:
I added a timeout to the join.